### PR TITLE
add a unit test for `convertToRaw`

### PR DESCRIPTION
**Pull request by mmusich**
#### PR description:

This is a follow-up to https://github.com/cms-sw/cmssw/pull/44735 in particular to the comment:

> (It would be good to add a unit test for this utility, but I currently do not have enough time to implement a proper one, and I would not delay this fix.)

add a basic unit test for the `convertToRaw` utility. 

#### PR validation:

`scram b runtests_testConvertToRaw` runs fine.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

N/A

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by missirol** [comment]
Thank you @mmusich !

File: HLTrigger/Tools/test/testConvertToRaw.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+echo
+
+convertToRaw --help  || die "Failure running convertToRaw --help" $?
+
+echo "========================================"
+echo "testing successful conversion"
+echo "----------------------------------------"
+echo
+
+check_for_success convertToRaw -f 1 -l=1 -v root://eoscms.cern.ch//eos/cms/store/data/Run2024C/EphemeralHLTPhysics0/RAW/v1/000/379/416/00000/e8dd5e3c-216f-4545-acb6-ab86c9161085.root
```
Comment (mmusich) [inline HLTrigger/Tools/test/testConvertToRaw.sh:34]: I am not sure if this is OK, but without the re-director `root://eoscms.cern.ch//eos/cms/` the test was failing: [log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-44a4e8/38926/unitTests/src/HLTrigger/Tools/test/testConvertToRaw/testing.log)

I would presume that once the file is cached we can remove the re-director. I don't know if there's a way of caching the file upfront (tagging @smuzaffar for suggestions - the file size is 900MB, hope we're ok with it).

Comment (iarspider) [reply HLTrigger/Tools/test/testConvertToRaw.sh:34]: @mmusich I think the size is OK (Shahzad can correct me if needed),and regarding caching this file - see similar discussion here: https://github.com/cms-sw/cmssw/issues/44293#issuecomment-1976849048 .

Comment (smuzaffar) [reply HLTrigger/Tools/test/testConvertToRaw.sh:34]: @mmusich  , this data file is now cached. Please use logical file name e.g. `convertToRaw -f 1 -l=1 -v /store/data/Run2024C/EphemeralHLTPhysics0/RAW/v1/000/379/416/00000/e8dd5e3c-216f-4545-acb6-ab86c9161085.root` . This should work for PR/IB tests. You can locally test it by setting ibeos cache paths
```
export CMS_PATH=/cvmfs/cms-ib.cern.ch
export SITECONFIG_PATH=/cvmfs/cms-ib.cern.ch/SITECONF/local
```

Comment (mmusich) [reply HLTrigger/Tools/test/testConvertToRaw.sh]: thank you. Done in the last push.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
so, the test fails with:

```
%MSG-w EvFDaqDirector:  PreGlobalEndLumi 18-Apr-2024 14:28:23 CEST  Run: 379416 Lumi: 97
 Handles to check for files to delete were not set by the input source...
%MSG
18-Apr-2024 14:28:23 CEST  Closed file root://eoscms.cern.ch//eos/cms/store/user/cmsbuild/store/data/Run2024C/EphemeralHLTPhysics0/RAW/v1/000/379/416/00000/e8dd5e3c-216f-4545-acb6-ab86c9161085.root
Traceback (most recent call last):
  File "/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_1_X_2024-04-18-1100/bin/el8_amd64_gcc12/convertToRaw", line 250, in <module>
    with open(eor_file) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_1_X_2024-04-18-1100/unit_tests/testConvertToRaw/run379416_ls0097/run379416/run379416_ls0000_EoR.jsn'
```

whereas on a `lxplus` node it works fine. Does the problem lie in the fact that the bot is not allowed to create sub-folders? Is there a way to test in an environment similar to the PR tests?

**Comment by smuzaffar** [comment]
please test

@mmusich ,  bot should have rights to create sub-dirs under `unit_tests/testConvertToRaw` . I have retsrated the tests and will monitor the build node

**Comment by smuzaffar** [comment]
@mmusich , running tests locally do generate `testConvertToRaw/run379416_ls0097/run379416/run379416_ls0000_EoR.jsn` file and for tests via bot I see that the generated file is `testConvertToRaw/run379416_ls0097/run379416/run000000_ls0000_EoR.jsn`. I really hav no idea why for bot tests the generated file has `run000000` instead of `run379416`

[a]
```
-rw-r--r--. 1 muzaffar zh 7108 Apr 22 09:40 ./testConvertToRaw.log
-rw-r--r--. 1 muzaffar zh  839 Apr 22 09:40 ./testConvertToRaw/run379416_cff.py
-rw-r--r--. 1 muzaffar zh    0 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/fu.lock
-rw-rw-rw-. 1 muzaffar zh  447 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/jsd/EoLS.jsd
-rw-rw-rw-. 1 muzaffar zh  439 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/jsd/EoR.jsd
-rw-rw-rw-. 1 muzaffar zh  128 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/jsd/rawData.jsd
-rwxrwxrwx. 1 muzaffar zh    0 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/open/fu-local.lock
-rw-rw-rw-. 1 muzaffar zh  213 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/run379416_ls0000_EoR.jsn
-rw-rw-rw-. 1 muzaffar zh  213 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/run379416_ls0097_EoLS.jsn
-rw-rw-rw-. 1 muzaffar zh  201 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/run379416_ls0097_index000000.jsn
-rwxrwxrwx. 1 muzaffar zh    0 Apr 22 09:40 ./testConvertToRaw/run379416_ls0097/run379416/run379416_ls0097_index000000.raw
```

[b]
```
-rw-rw-rw- 1 cmsbld cmsbld    447 Apr 22 10:05 ./testConvertToRaw/run379416/jsd/EoLS.jsd
-rw-rw-rw- 1 cmsbld cmsbld    439 Apr 22 10:05 ./testConvertToRaw/run379416/jsd/EoR.jsd
-rw-rw-rw- 1 cmsbld cmsbld    128 Apr 22 10:05 ./testConvertToRaw/run379416/jsd/rawData.jsd
-rw-rw-rw- 1 cmsbld cmsbld    236 Apr 22 10:05 ./testConvertToRaw/run379416/run379416_ls0097_index000000.jsn
-rwxrwxrwx 1 cmsbld cmsbld 415408 Apr 22 10:05 ./testConvertToRaw/run379416/run379416_ls0097_index000000.raw
-rw-r--r-- 1 cmsbld cmsbld      0 Apr 22 10:05 ./testConvertToRaw/run379416_ls0097/run379416/fu.lock
-rwxrwxrwx 1 cmsbld cmsbld      0 Apr 22 10:05 ./testConvertToRaw/run379416_ls0097/run379416/open/fu-local.lock
-rw-rw-rw- 1 cmsbld cmsbld    248 Apr 22 10:05 .testConvertToRaw/run379416_ls0097/run379416/run000000_ls0000_EoR.jsn
-rw-rw-rw- 1 cmsbld cmsbld    248 Apr 22 10:05 ./testConvertToRaw/run379416_ls0097/run379416/run000000_ls0097_EoLS.jsn
```

**Comment by smuzaffar** [comment]
I think problem is `https://github.com/cms-sw/cmssw/blob/3fb8b0edb88bcba34579ca4b9867427862eda51d/EventFilter/Utilities/plugins/RawEventFileWriterForBU.cc#L312` . It tries to search for `run` and during PR tests it found run from the build path `/data/cmsbld/jenkins/workspace/ib-run-pr-tests`

**Comment by mmusich** [comment]
> It tries to search for run and during PR tests it found run from the build path 

good catch!
Perhaps something like:
```diff
diff --git a/EventFilter/Utilities/plugins/RawEventFileWriterForBU.cc b/EventFilter/Utilities/plugins/RawEventFileWriterForBU.cc
index ccfd7b51b5a..ed512b78d42 100644
--- a/EventFilter/Utilities/plugins/RawEventFileWriterForBU.cc
+++ b/EventFilter/Utilities/plugins/RawEventFileWriterForBU.cc
@@ -309,8 +309,8 @@ void RawEventFileWriterForBU::stop() {
 //TODO:get from DaqDirector !
 void RawEventFileWriterForBU::makeRunPrefix(std::string const& destinationDir) {
   //dirty hack: extract run number from destination directory
-  std::string::size_type pos = destinationDir.find("run");
-  std::string run = destinationDir.substr(pos + 3);
+  std::string::size_type pos = destinationDir.find("/run");
+  std::string run = destinationDir.substr(pos + 4);
   run_ = atoi(run.c_str());
   std::stringstream ss;
   ss << "run" << std::setfill('0') << std::setw(6) << run_;
```

it's enough, but I am wary of making this change to DAQ core code without explicit green-light.

**Comment by smuzaffar** [comment]
`std::string::size_type pos = destinationDir.find("/run");` can still fail if `destinationDir` contains `/run` in it. A bit safe would be to use `rfind` e.g. `std::string::size_type pos = destinationDir.rfind("/run");`

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by smuzaffar** [comment]
By the way , @@smorovic, the only place [RawEventFileWriterForBU::initialize(....)](https://github.com/cms-sw/cmssw/blob/master/EventFilter/Utilities/plugins/RawEventFileWriterForBU.cc#L104) is called is from https://github.com/cms-sw/cmssw/blob/master/EventFilter/Utilities/plugins/RawEventOutputModuleForBU.h . So why change not `RawEventFileWriterForBU::initialize()` signature and pass `edm::Service<evf::EvFDaqDirector>()->getRunNumber()` to it instead of extrating the run number from the `destinationDir`

**Comment by smorovic** [comment]
> By the way , @@smorovic, the only place [RawEventFileWriterForBU::initialize(....)](https://github.com/cms-sw/cmssw/blob/master/EventFilter/Utilities/plugins/RawEventFileWriterForBU.cc#L104) is called is from https://github.com/cms-sw/cmssw/blob/master/EventFilter/Utilities/plugins/RawEventOutputModuleForBU.h . So why change not `RawEventFileWriterForBU::initialize()` signature and pass `edm::Service<evf::EvFDaqDirector>()->getRunNumber()` to it instead of extrating the run number from the `destinationDir`

I agree, it could also be done in this way (DaqDirector should know run number as well, even in "BU" mode).

**Comment by mmusich** [comment]
> I agree, it could also be done in this way (DaqDirector should know run number as well, even in "BU" mode).

I would leave that to a further update from DAQ experts and make this PR as the minimal changeset to allow running the new unit test.

**Comment by smorovic** [comment]
+1

**Comment by mmusich** [comment]
+hlt

**Comment by rappoccio** [comment]
+1

**Pull request merged**